src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -1754,6 +1754,11 @@ protected internal virtual TypeDefinition MarkType (TypeReference reference, Dep
 			foreach (Action<TypeDefinition> handleMarkType in _markContext.MarkTypeActions)
 				handleMarkType (type);
 
+			if (type.BaseType != null &&
+				!_context.Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (type) &&
+				_context.Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (_context.TryResolve (type.BaseType)))
```

Comment:
I think this check needs to look for RUC not just on the type, but on declaring types in case they are nested. For example:

```csharp
[RUC]
public class Outer {
  public class Base {}
}

class Derived : Outer.Base {} // should warn
```

Reply:
I think it's slightly trickier than that if we want to follow Obsolete:

```C#
using System;
[Obsolete]
public class C {
    public class D { }
    public void M() {
        var d = new D(); // no warning
    }
}

class E
{
    public static void M()
    {
        var d = new C.D(); // warning
    }
}
```

Reply:
I think we should use Obsolete as a good way to tell what is a good UX - and then ask if it's something which is reasonable to achieve in the linker/analyzer and if it's worth it.

This example from @agocke should already work that way, because the first call (where it should not warn) will be suppressed by the RUC being on the parent type.

Reply:
For this example, where do you expect to see the warning?

```csharp
[RUC]
public class Outer {
  public class Base {}
}

class Derived : Outer.Base {} // should warn
```

I would expect to see one for the Derived's constructor because it's calling the constructor on Outer.Base. Does that match your expectations?

Reply:
I would expect to see a warning that `Derived` derives from `Outer.Base` which is RUC. And ideally no other warnings (byt we might need to do some work to not generate more).

Reply:
Now it warns in the following way
```C#
[RUC]
public class Outer {
  public class Base {}
}

// IL2109: Type 'Derived' derives from 'Outer.Base' which has 'RequiresUnreferencedCodeAttribute'.
class Derived : Outer.Base {}
```